### PR TITLE
Update to transformers 2.3.0 & Add ALBERT

### PR DESCRIPTION
**Pull request by HaokunLiu**
### Important notice
You will need to change PYTORCH_PRETRAINED_BERT_CACHE to HUGGINGFACE_TRANSFORMERS_CACHE in your own env setting.

### Updates
This PR update from pytorch_transformers 1.0 to transformers 2.3. The major changes are
1. Add albert to input modules
2. Rename mentions of "pytorch_pretrained_bert", "pytorch_transformers" to "transformers". Sometimes the name "huggingface transformers" is used to avoid ambiguity. I only wish they do not transform their name anymore.
3. As #953 pointed out, due to huggingface reverted their bytebpe tokenizer, our temporary fix does not work. This PR changes the alignment function in qa.py back to using TokenAligner.

### Related Issues
#972  #920  #730

### Other
Transformers 2.3 introduced a new feature: AutoModel, AutoTokenizer etc, this may be used to simplify some code. But it is not very necessary right now, so I didn't do it.

### CODE REVIEW

**Comment by HaokunLiu** [comment]
So it turned out ALBERT really is better than RoBERTa. I run some experiments on cola and rte.
| task | roberta-large | albert-xxlarge-v1 | albert-xxlarge-v2 |
|------|---------------|-------------------|-------------------|
| cola | 0.685         | 0.695             | 0.701             |
| rte  | 0.862         | 0.888             | 0.906             |
I didn't use the exact same hyperparameter. But the result is already close to their dev set results on the paper.

**Comment by sleepinyourhat** [comment]
Great! How hard was it to run albert-xxlarge? Did you have to use smaller
batch sizes than for roberta-large?

**Comment by HaokunLiu** [comment]
> Great! How hard was it to run albert-xxlarge? Did you have to use smaller batch sizes than for roberta-large?
> […](#)
> On Wed, Jan 15, 2020 at 10:00 AM Haokun Liu ***@***.***> wrote: So it turned out ALBERT really is better than RoBERTa. I run some experiments on cola and rte. | task | roberta-large | albert-xxlarge-v1 | albert-xxlarge-v2 | |------|---------------|-------------------|-------------------| | cola | 0.685 | 0.695 | 0.701 | | rte | 0.862 | 0.888 | 0.906 | I didn't use the exact same hyperparameter. But the result is already close to their dev set results on the paper. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_nyu-2Dmll_jiant_pull_990-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAAJZSWOPL5K6BC6ERCBS5YDQ54QHHA5CNFSM4KFPSL5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJATEHI-23issuecomment-2D574698013&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=sCzLyHdE8zgQwk2-sKwA1w&m=ACW_26To1ILze-7xVTqVDQT-Y1fafHMNQGxlpSAqBr8&s=yaMZ4q5CTG74o8g5OYHqmv-vRfh8VRitn61kws6aCJg&e=>, or unsubscribe <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAJZSWJ5DGQCPY56PJMWKLTQ54QHHANCNFSM4KFPSL5A&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=sCzLyHdE8zgQwk2-sKwA1w&m=ACW_26To1ILze-7xVTqVDQT-Y1fafHMNQGxlpSAqBr8&s=0F--szix3dfGhsC8TOf1VuzWD7yhqS9PzmkRoaaMLoA&e=> .

Input of RTE and CoLA are fairly short, it does not cause any memory problem. But since albert-xxlarge have 4x the hidden size of roberta-large, and half number of layers. It will have to use smaller batch sizes, if the task has greater input length.

**Comment by HaokunLiu** [comment]
Comparing RoBERTa tokenizer in pytorch_transformers ('ĠBerlin', 'Ġand', 'ĠMunich') and RoBERTa tokenizer in transformers ('Ber', 'lin', 'Ġand', 'ĠMunich') on QAMR.

| pytorch_transformers | transformers |
| --- | --- |
|0.730| 0.726 |

Yes, the new one seems very counter-intuitive, but that's what Huggingface finally settled on. Some people found the new one marginally better on NER . https://github.com/huggingface/transformers/issues/1196

**Comment by sleepinyourhat** [comment]
Taking a look now...

File: jiant/huggingface_transformers_interface/modules.py [hunk 1]
```diff
@@ -355,26 +354,83 @@ def forward(self, sent: Dict[str, torch.LongTensor], task_name: str = "") -> tor
...
-        model_with_lm_head = pytorch_transformers.RobertaForMaskedLM.from_pretrained(
+        model_with_lm_head = transformers.RobertaForMaskedLM.from_pretrained(
             self.input_module, cache_dir=self.cache_dir
         )
         lm_head = model_with_lm_head.lm_head
         lm_head.predictions.decoder.weight = self.model.embeddings.word_embeddings.weight
         return nn.Sequential(lm_head, nn.LogSoftmax(dim=-1))
 
 
-class XLNetEmbedderModule(PytorchTransformersEmbedderModule):
```
Comment (sleepinyourhat) [inline jiant/huggingface_transformers_interface/modules.py]: This all looks reasonable, though it's frustrating that we can't reuse more of this code.

Could you highlight the pieces of this that you had to write yourself?

File: jiant/huggingface_transformers_interface/modules.py [hunk 2]
```diff
@@ -355,26 +354,83 @@ def forward(self, sent: Dict[str, torch.LongTensor], task_name: str = "") -> tor
...
+        )
+        self.max_pos = self.model.config.max_position_embeddings
+
+        self.tokenizer = transformers.AlbertTokenizer.from_pretrained(
+            args.input_module, cache_dir=self.cache_dir
+        )  # TODO: Speed things up slightly by reusing the previously-loaded tokenizer.
+        self._sep_id = self.tokenizer.convert_tokens_to_ids("[SEP]")
+        self._cls_id = self.tokenizer.convert_tokens_to_ids("[CLS]")
+        self._pad_id = self.tokenizer.convert_tokens_to_ids("[PAD]")
+        self._unk_id = self.tokenizer.convert_tokens_to_ids("[UNK]")
```
Comment (sleepinyourhat) [inline jiant/huggingface_transformers_interface/modules.py]: Double check these. I see Huggingface uses `<unk>` in a couple of places:
https://github.com/huggingface/transformers/pull/1683/files#diff-78dc6baf03f4d704be366101942aadd3R62

File: jiant/utils/retokenize.py [hunk 3]
```diff
@@ -95,35 +95,6 @@ def _mat_from_spans_sparse(spans: Sequence[Tuple[int, int]], n_chars: int) -> Ma
     return sparse.csr_matrix((data, (ridxs, cidxs)), shape=(len(spans), n_chars))
```
Comment (sleepinyourhat) [inline jiant/utils/retokenize.py]: I don't really know the logic in this file. You'll need somebody else to take a look at it.

File: setup.py [hunk 4]
```diff
@@ -65,7 +65,7 @@
         "pyhocon==0.3.35",
         "python-Levenshtein==0.12.0",
         "sacremoses",
```
Comment (sleepinyourhat) [inline setup.py]: This was deleted above.

File: environment.yml [hunk 5]
```diff
@@ -39,9 +39,8 @@ dependencies:
         - sacremoses
 
         # Warning: jiant currently depends on *both* pytorch_pretrained_bert > 0.6 _and_
-        # pytorch_transformers > 1.0. These are the same package, though the name changed between
+        # transformers > 2.3.0. These are the same package, though the name changed between
         # these two versions. AllenNLP requires 0.6 to support the BertAdam optimizer, and jiant
         # directly requires 1.0 to support XLNet and WWM-BERT.
         # This AllenNLP issue is relevant: https://github.com/allenai/allennlp/issues/3067
-        - sacremoses
```
[the remark this answers is not in the document]
Comment (HaokunLiu) [reply environment.yml]: this line is duplicated, we already asked for sacremoses in line 39

File: jiant/config/defaults.conf [hunk 6]
```diff
@@ -284,16 +287,16 @@ d_char = 100  // Dimension of trained char embeddings.
 n_char_filters = 100  // Number of filters in trained char CNN.
 char_filter_sizes = "2,3,4,5"  // Size of char CNN filters.
 
-pytorch_transformers_output_mode = "none"  // How to handle the embedding layer of the
+transformers_output_mode = "none"  // How to handle the embedding layer of the
                                            // BERT/XLNet model:
```
[the remark this answers is not in the document]
Comment (HaokunLiu) [reply jiant/config/defaults.conf]: done

File: jiant/config/defaults.conf [hunk 7]
```diff
@@ -244,20 +244,23 @@ input_module = ""  // The word embedding or contextual word representation layer
...
                    //       ELMo, but not ELMo's LSTM layer hidden states. Use with
                    //       tokenizer = MosesTokenizer.
-                   //   - bert-base-uncased, etc.: Any BERT model from pytorch_transformers.
+                   //   - bert-base-uncased, etc.: Any BERT model from transformers.
                    //   - roberta-base / roberta-large / roberta-large-mnli: RoBERTa model from
-                   //       pytorch_transformers.
+                   //       transformers.
+                   //   - albert-base-v1 / albert-large-v1 / albert-xlarge-v1 / albert-xxlarge-v1
+                   //   - albert-base-v2 / albert-large-v2 / albert-xlarge-v2 / albert-xxlarge-v2:
+                   //       ALBERT model encoder from transformers.
```
[the remark this answers is not in the document]
Comment (HaokunLiu) [reply jiant/config/defaults.conf]: done

File: jiant/huggingface_transformers_interface/__init__.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+These are the same package, though the name changed between these two versions. AllenNLP requires
+0.6 to support the BertAdam optimizer, and jiant directly requires 2.3.
+
+This AllenNLP issue is relevant: https://github.com/allenai/allennlp/issues/3067
+
+TODO: We do not support non-English versions of XLM, if you need them, add some code in XLMEmbedderModule
+to prepare langs input to transformers.XLMModel
+"""
+
+# All the supported input_module from huggingface transformer
```
[the remark this answers is not in the document]
Comment (HaokunLiu) [reply jiant/huggingface_transformers_interface/__init__.py]: thanks

File: jiant/huggingface_transformers_interface/__init__.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+# input_modules mapped to the same string share vocabulary
```
[the remark this answers is not in the document]
Comment (HaokunLiu) [reply jiant/huggingface_transformers_interface/__init__.py]: e.g. bert-base-uncased and bert-large-uncased have the same tokenizer, wordpiece with the exact same vocabulary in it. So, cache files like preprocessed data under the experiment directory can be shared across runs, even if one uses bert-base-uncased and the other uses bert-large-uncased. Unfortunately, at this moment, we don't have this level of cache file management in jiant (problems with cache files actually cause a lot of confusion and reruns in projects of our own and external users).

[the remark this answers is not in the document]
Comment (HaokunLiu) [reply jiant/huggingface_transformers_interface/__init__.py]: I put this dictionary here so that if someone else want to build that kind system, it will be relatively easy.

File: jiant/utils/retokenize.py (hunk 3 above)
Comment (HaokunLiu) [reply jiant/utils/retokenize.py]: I think @iftenney  know this the best. @pruksmhc  and @zphang also know about it.

File: setup.py (hunk 4 above)
Comment (HaokunLiu) [reply setup.py]: We still need this. There were two lines of this in the file above.

File: environment.yml [hunk 10]
```diff
@@ -39,9 +39,8 @@ dependencies:
         - sacremoses
 
         # Warning: jiant currently depends on *both* pytorch_pretrained_bert > 0.6 _and_
```
[the remark this answers is not in the document]
Comment (HaokunLiu) [reply environment.yml]: At this point, I'm not familiar with AllenNLP enough to do that well. Maybe it's better to bundle this issue with #657 .

File: jiant/huggingface_transformers_interface/modules.py (hunk 2 above)
Comment (HaokunLiu) [reply jiant/huggingface_transformers_interface/modules.py]: Thanks, you are right. They have [SEP] [CLS] <pad> and <unk>.

Comment (sleepinyourhat) [reply jiant/huggingface_transformers_interface/modules.py]: Do triple check that section of code. Errors like this could hurt performance in a subtle and hard-to-diagnose way across lots of examples.

**Comment by pyeres** [comment]
> This all looks reasonable, but for a couple of minor points. This is big enough that I'd like somebody else to take a look before we merge, though. Especially since this touches retokenization logic that I haven't really paid much attention to so far. @pyeres @pruksmhc, could either of you give the main review?
> 
> The overall diff is huge, but I don't see an easy way of factoring it, especially since most of the line count lives in the variable name changes. That said, if anyone sees a clean split, we should split this up.

@HaokunLiu, I'm starting to review now. Here are the validation items we discussed:
- [ ] Performance regression test (performance on some core tasks not reduced w/ these changes)
- [ ] Benchmark test — Albert performance in jiant is comparable to paper
- [ ] Test coverage on tokenization code

File: jiant/huggingface_transformers_interface/__init__.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+This AllenNLP issue is relevant: https://github.com/allenai/allennlp/issues/3067
+
+TODO: We do not support non-English versions of XLM, if you need them, add some code in XLMEmbedderModule
+to prepare langs input to transformers.XLMModel
+"""
+
+# All the supported input_module from huggingface transformers
+# input_modules mapped to the same string share vocabulary
+input_module_to_pretokenized = {
```
Comment (pyeres) [inline jiant/huggingface_transformers_interface/__init__.py]: input_module_to_pretokenized -> transformer_input_module_to_tokenizer_id

File: jiant/huggingface_transformers_interface/modules.py (hunk 1 above)
Comment (HaokunLiu) [reply jiant/huggingface_transformers_interface/modules.py]: Transformers differs slightly in what are the special tokens, where those special tokens are added in the input sequence, some input output interface. While their embedding modules and  language model heads can sometimes have considerable differences.
I know Huggingface has code that can swap in&out different kind of transformers, and their out most interface is shared (sometimes by using dummy optional inputs, which is confusing). But since we need to support things like using only word embedding and mixing inner representation, we often have to go into each transformer model and dealing with their internal differences.
Overall, I think there isn't much space for further simplification without sacrificing readability.

File: environment.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (pruksmhc) [reply environment.yml]: I'm looking into upgrading AllenNLP to integrate their >512 sliding window approach.

**Comment by HaokunLiu** [comment]
Actually, don't merge this now. I found our current implementation of many tasks use the assumption that tokenizing each word independently and concatenating them will give you the same result as tokenizing the full sentence. The tasks may include CCG, ReCoRD, WiC. I need to run some more tests, and possibly make some changes to these tasks.

**Comment by HaokunLiu** [comment]
Task | Update to transformers | Final baseline in Taskmaster | Initial hyper-parameter search in Taskmaster
-- | -- | -- | --
Target | |
commitbank | 1 | 0.991 | 1
copa | 0.94 | 0.86 | 0.93
wsc | 0.692 | 0.673 | 0.798
rte | 0.866 | 0.835 | 0.862
multirc | 0.628 | 0.474 | 0.574
wic | 0.724 | 0.705 | 0.721
boolq | 0.879 | 0.866 | 0.877
commonsenseqa | 0.735 | 0.74 | 0.734
cosmosqa | 0.814 | 0.819 | 0.817
record | 0.863 | 0.86 | 0.898
INTERM |   |  
qamr | 0.726 | N/A | 0.735
scitail | 0.976 | N/A | 0.972
social-iqa | 0.766 | N/A | 0.729
ccg | 0.96 | N/A | 0.96
hellaswag | 0.851 | N/A | 0.856
qasrl | 0.664 | N/A | 0.675
sst | 0.961 | N/A | 0.968
qqp | 0.903 | N/A | 0.898
mnli | 0.896 | N/A | 0.901

The first column is the updated roberta-large model, all the tasks use lr=5e-6 and dropout=0.2, the second column is the average result from three random seeds using the previously found "optimal" learning rate and dropout rate, the third column is the best result we got when doing hyper-parameter searching.

Most results are on par with our previous result. Some are marginally lower, I think it's understandable, since it's just using a single hyper-parameter setting. Except for WSC, which seems very unstable.

**Comment by pyeres** [comment]
> Most results are on par with our previous result. Some are marginally lower, I think it's understandable, since it's just using a single hyper-parameter setting. Except for WSC, which seems very unstable.

This looks OK to me — on tasks where performance with this updated transformers code is below the level reported in your "Initial hyper-parameter search in Taskmaster" column, performance with your updated transformers code is above (or very close to) the performance in your "Final baseline in Taskmaster" column (and I understand that these final baselines are the result of multiple runs).

**Comment by pyeres** [comment]
Thanks @HaokunLiu for running the additional performance validations. Merging.
